tsl/src/nodes/decompress_chunk/exec.c

```diff
@@ -431,346 +277,240 @@ decompress_chunk_begin(CustomScanState *node, EState *estate, int eflags)
-				 */
-				AttrNumber attr = AttrNumberGetAttrOffset(column->output_attno);
-				batch_state->decompressed_slot_scan->tts_values[attr] =
-					slot_getattr(batch_state->compressed_slot,
-								 column->compressed_scan_attno,
-								 &batch_state->decompressed_slot_scan->tts_isnull[attr]);
-				break;
-			}
-			case COUNT_COLUMN:
+			DecompressChunkColumnDescription *column = &chunk_state->template_columns[i];
+			if (column->bulk_decompression_supported)
 			{
-				value = slot_getattr(batch_state->compressed_slot,
-									 column->compressed_scan_attno,
-									 &isnull);
-				/* count column should never be NULL */
-				Assert(!isnull);
-				int count_value = DatumGetInt32(value);
-				if (count_value <= 0)
-				{
-					ereport(ERROR,
-							(errmsg("the compressed data is corrupt: got a segment with length %d",
-									count_value)));
-				}
-
-				if (batch_state->total_batch_rows == 0)
-				{
-					batch_state->total_batch_rows = count_value;
-				}
-				else if (batch_state->total_batch_rows != count_value)
-				{
-					elog(ERROR, "compressed column out of sync with batch counter");
-				}
-
-				break;
+				/* Values array, with 64 element padding (actually we have less). */
+				chunk_state->batch_memory_context_bytes +=
+					(GLOBAL_MAX_ROWS_PER_COMPRESSION + 64) * column->value_bytes;
+				/* Also nulls bitmap. */
+				chunk_state->batch_memory_context_bytes +=
+					GLOBAL_MAX_ROWS_PER_COMPRESSION / (64 * sizeof(uint64));
+				/* Arrow data structure. */
+				chunk_state->batch_memory_context_bytes +=
+					sizeof(ArrowArray) + sizeof(void *) * 2 /* buffers */;
+				/* Memory context header overhead for the above parts. */
+				chunk_state->batch_memory_context_bytes += sizeof(void *) * 3;
 			}
-			case SEQUENCE_NUM_COLUMN:
-				/*
-				 * nothing to do here for sequence number
-				 * we only needed this for sorting in node below
-				 */
-				break;
 		}
 	}
-	batch_state->initialized = true;
-	MemoryContextSwitchTo(old_context);
-}
 
-/* Perform the projection and selection of the decompressed tuple */
-static bool pg_nodiscard
-decompress_chunk_perform_select_project(CustomScanState *node,
-										TupleTableSlot *decompressed_slot_scan,
-										TupleTableSlot *decompressed_slot_projected)
-{
-	ExprContext *econtext = node->ss.ps.ps_ExprContext;
+	/* Round up to even number of 4k pages. */
+	chunk_state->batch_memory_context_bytes =
+		((chunk_state->batch_memory_context_bytes + 4095) / 4096) * 4096;
 
-	/*
-	 * Reset expression memory context to clean out any cruft from
-	 * previous batch. Our batches are 1000 rows max, and this memory
-	 * context is used by ExecProject and ExecQual, which shouldn't
-	 * leak too much. So we only do this per batch and not per tuple to
-	 * save some CPU.
-	 */
-	econtext->ecxt_scantuple = decompressed_slot_scan;
-	ResetExprContext(econtext);
+	/* As a precaution, limit it to 1MB. */
+	chunk_state->batch_memory_context_bytes =
```

Comment:
Could we output the chosen memory size in a debug message? This would allow us to see in a production instance which value was selected.

Reply:
Added a message.